db/migrate/20230121005809_add_rubygem_to_position_versions_index.rb

```diff
@@ -0,0 +1,11 @@
+class AddRubygemToPositionVersionsIndex < ActiveRecord::Migration[7.0]
+  def up
+    add_index :versions, [:position, :rubygem_id]
```

Comment:
This migration looks like it will lock the `versions` table when we run `db:migrate`, can we verify if it is or isn't and how we can run the index migration in parallel to avoid locking?

Reply:
Yes, my plan is to deploy this manually using concurrently.

1. create new index with temporary name (using CONCURRENTLY)
2. rename old index
3. rename new index
4. drop old index

Reply:
Could we not do that inside this migration? it would be good to see & review the exact steps we're planning to perform in production.

Reply:
I was about to keep it simple in the code. Otherwise we will need to use `disable_ddl_transaction!` which makes it unfriendly to automated deploy (since it can fail at any point and the result is not predictable in "fail state").

In theory we can run it as is since the index could be created within 2 seconds (tested on staging). And versions (if I understand it well) will be blocking only push/yank for the given amount of time.

The manual deployment as described is just to make it super smooth (and is related to manual deploy since you can simply check all steps are properly executed). Automating the same doesn't make sense to me. And the final state is the same, so it makes really no difference.

Also this migration could be deployed ahead of the code. It works well for the old code and also for the new one. Deploying it manually before deploy makes it even smoother.

Personally I have good long-term experience with this approach to make DB migrations smooth as possible.

If you still prefer to write down the SQL into the migration exactly as is going to be executed, I can do that. I will still prefer to run it manually before the deploy.

---

_I have checked the other migrations and there is only one similar (https://github.com/rubygems/rubygems.org/blob/master/db/migrate/20200429005140_add_dash_underscore_typo_detection_index_to_rubygems.rb) written in a little dangerous way (there is small time in between DROP and CREATE when system is without the index)._

Reply:
Btw. it seems generated index names are different, the plan is even easier. I can leave as a comment in well in the migration.

```sql
CREATE INDEX CONCURRENTLY index_versions_on_position_and_rubygem_id ON versions USING btree(position, rubygem_id);
DROP INDEX CONCURRENTLY index_versions_on_position;
INSERT INTO schema_migrations VALUES (('20230121005809'));

```

Reply:
Thanks for the additional context! With the reasons you mentioned, I prefer to keep the migration it as is then, and leave a comment of what you deployed manually.

Reply:
I have left comment in the migration @jenshenny.